server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/SHARDS.java

```diff
@@ -0,0 +1,255 @@
+
+   public SHARDS() {
+      super(2, 0, 0, 0);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      AdvancedCache<?, ?> respCache = handler.cache();
+      DistributionManager dm = respCache.getDistributionManager();
+      if (dm == null) {
+         Configuration configuration = SecurityActions.getCacheConfiguration(respCache);
+         int segments = configuration.clustering().hash().numSegments();
+         NodeInformation local = readLocalNodeInformation(respCache.getCacheManager());
+         List<IntRange> ranges = List.of(new IntRange(0, segments - 1));
+         String res = serialize(List.of(new RangeOwners(local, null, ranges)));
+         ByteBufferUtils.stringToByteBuf(res, handler.allocatorToUse());
+         return handler.myStage();
+      }
+
+      CacheTopology topology = dm.getCacheTopology();
+      ConsistentHash hash = topology.getCurrentCH();
+
+      if (hash == null) {
+         return NO_CONSISTENT_HASH;
+      }
+
+      return handler.stageToReturn(readShardsInformation(hash, SecurityActions.getClusterExecutor(respCache)), ctx, ByteBufferUtils::stringToByteBuf);
+   }
+
+   private static CompletionStage<String> readShardsInformation(ConsistentHash hash, ClusterExecutor executor) {
+      Map<List<Address>, Set<Integer>> segmentOwners = new HashMap<>();
+      for (int i = 0; i < hash.getNumSegments(); i++) {
+         segmentOwners.computeIfAbsent(hash.locateOwnersForSegment(i), ignore -> new HashSet<>())
+               .add(i);
+      }
+
+      return readNodeInformation(hash.getMembers(), executor)
+            .thenApply(information -> {
+               List<RangeOwners> response = new ArrayList<>();
+               for (Map.Entry<List<Address>, Set<Integer>> entry : segmentOwners.entrySet()) {
+                  List<Address> addresses = entry.getKey();
+                  List<IntRange> ranges = findRanges(entry.getValue());
+                  NodeInformation leader = information.get(addresses.get(0));
+                  if (leader == null) {
+                     throw new IllegalStateException("Unknown leader while collecting cluster information");
+                  }
+                  List<NodeInformation> replicas = null;
+                  if (addresses.size() > 1) {
+                     replicas = new ArrayList<>();
+                     for (int i = 1; i < addresses.size(); i++) {
+                        replicas.add(information.get(addresses.get(i)));
+                     }
+                  }
+                  response.add(new RangeOwners(leader, replicas, ranges));
+               }
+               return serialize(response);
+            });
+   }
+
+   private static CompletionStage<Map<Address, NodeInformation>> readNodeInformation(List<Address> members, ClusterExecutor executor) {
+      final Map<Address, NodeInformation> responses = new ConcurrentHashMap<>(members.size());
```

Comment:
Do we still not store the NodeInformation for each node in each node yet...? This should be easy to keep up to date, as it only needs an update when a new node joins or leaves.

Reply:
Added a little state in this command. If the consistent hash still equals, we return the previous execution.